pytorch_lightning/trainer/trainer.py

```diff
@@ -856,12 +856,13 @@ def _run(self, model: 'pl.LightningModule') -> Optional[Union[_EVALUATE_OUTPUT,
         # restore modules after setup
         self.checkpoint_connector.restore_datamodule()
         self.checkpoint_connector.restore_model()
-        # restore callback states
-        self.checkpoint_connector.restore_callbacks()
 
         self._call_configure_sharded_model(model)  # allow user to setup in model sharded environment
         self.accelerator.setup(self, model)  # note: this sets up self.lightning_module
 
+        # restore callback states
+        self.checkpoint_connector.restore_callbacks()
+
```

Comment:
it was changed here: #7652

was the reported issue working fine before this?
ideally, we want to pair `restore_model()` and `restore_callbacks()` together such that the on_load_checkpoint runs right after each other. 

now there are additional hooks running between.
is it not possible to delay optimizer setup in the fine tuning callbacks to `on_train_start` or similar?

Reply:
Hey @awaelchli, the test should fail on master due to position of the `setup`. However, if the user was using sharded optimizers, we would need to capture them too.

Reply:
@awaelchli The notebook from the issue https://github.com/PyTorchLightning/pytorch-lightning/issues/8502 ([here](https://colab.research.google.com/drive/1xIE218xOvK9tSoWV_2z820Qog_NqrILs?usp=sharing)) works with version 1.3.8.

Reply:
Done !